### PR TITLE
[FLINK-10945] Use InputDependencyConstraint to avoid resource dead…

### PR DESCRIPTION
**Pull request by zhuzhurk**
…locks in LAZY_FROM_SOURCES scheduling when resources are limited

## What is the purpose of the change

This PR add a job config InputDependencyConstraint, which helps to avoid resource deadlocks in LAZY_FROM_SOURCES scheduling when resources are limited, as described in [FLINK-10945](https://issues.apache.org/jira/browse/FLINK-10945).

## Brief change log

 - *Add InputDependencyConstraint to ExecutionConfig*
 - *Adjust isConsumable interface in IntermediateResultPartition to fit for the data actual consumable definition*
 - Change current execution lazy scheduling logic(in Execution.scheduleOrUpdateConsumers(edges)) to schedule tasks only if the InputDependencyConstraint is satisfied(an interface ExecutionVertex.checkInputDependencyConstraints is added to serve this purpose). 

## Verifying this change

This change added tests and can be verified as follows:

  - *Added IntermediateResultPartitionTest to validate IntermediateResultPartition changes*
  - *Added ExecutionVertexInputConstraintTest to validate the constraint check logic in ExecutionVertex*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

File: flink-runtime/src/test/java/org/apache/flink/runtime/deployment/InputChannelDeploymentDescriptorTest.java [hunk 1]
```diff
@@ -192,6 +193,7 @@ private static ExecutionVertex mockExecutionVertex(ExecutionState state, Resourc
 
 	private static IntermediateResultPartition mockPartition(ExecutionVertex producer) {
 		IntermediateResultPartition partition = mock(IntermediateResultPartition.class);
+		when(partition.getResultType()).thenReturn(ResultPartitionType.PIPELINED);
```
Comment (azagrebin) [inline flink-runtime/src/test/java/org/apache/flink/runtime/deployment/InputChannelDeploymentDescriptorTest.java]: Is this change needed? `isConsumable` should be still enough.

Comment (zhuzhurk) [reply flink-runtime/src/test/java/org/apache/flink/runtime/deployment/InputChannelDeploymentDescriptorTest.java]: Yes. It is needed to avoid NPE in the tests.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/IntermediateResultPartition.java [hunk 2]
```diff
@@ -36,6 +36,11 @@
 
 	private List<List<ExecutionEdge>> consumers;
 
+	/**
+	 * Whether this partition has data produced. For pipelined result only.
+	 */
+	private boolean dataProduced = false;
```
[the remark this answers is not in the document]
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/IntermediateResultPartition.java]: That's fine.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java [hunk 3]
```diff
@@ -747,31 +747,34 @@ else if (numConsumers == 0) {
...
+				if (consumerVertex.checkInputDependencyConstraints()) {
+					// When deploying a consuming task, its task deployment descriptor will contain all
+					// deployment information available at the respective time. It is possible that some
+					// of the partitions to be consumed have not been created yet. These are updated
+					// runtime via the update messages.
+					//
+					// TODO The current approach may send many update messages even though the consuming
+					// task has already been deployed with all necessary information. We have to check
+					// whether this is a problem and fix it, if it is.
+					CompletableFuture.supplyAsync(
```
[the remark this answers is not in the document]
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java]: Sure.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/IntermediateResultPartition.java [hunk 4]
```diff
@@ -36,6 +36,11 @@
 
 	private List<List<ExecutionEdge>> consumers;
 
+	/**
+	 * Whether this partition has data produced. For pipelined result only.
```
[the remark this answers is not in the document]
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/IntermediateResultPartition.java]: Agree.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java [hunk 5]
```diff
@@ -726,6 +728,56 @@ void sendPartitionInfos() {
 		}
 	}
 
+	/**
+	 * Check whether the InputDependencyConstraint is satisfied for this vertex.
+	 *
+	 * @return whether the input constraint is satisfied
+	 */
+	public boolean checkInputDependencyConstraints() {
```
[the remark this answers is not in the document]
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java]: Thanks for the suggestion. I made it more concise, but a bit different from your sample code:
For PIPELINED input consumable check, only the partitions related with the input edges are checked. 

I'm not sure why you think the behavior is changed? 
I'd agree if you have concerns that this check could be heavy(O(N) complexity)) for large parallelism vertex. So I'm considering whether to add a short cut check for `InputDependencyConstraint.ANY` in `Execution.scheduleOrUpdateConsumers`, i.e. change `if (consumerVertex.checkInputDependencyConstraints()) ` to `if (getVertex().getExecutionGraph().getInputDependencyConstraint() == InputDependencyConstraint.ANY || consumerVertex.checkInputDependencyConstraints())`. It's not elegant but can be as efficient as current scheduling.

The input state is not used in EAGER scheduling. So the InputDependencyConstraint is not a matter in EAGER mode. It is used in scheduleOrUpdateConsumers, which is only invoked in LAZY_FROM_SOURCES scheduling.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionVertexInputConstraintTest.java [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+		ev11.fail(new Exception());
+		waitUntilJobRestarted(eg);
+		assertFalse(ev31.checkInputDependencyConstraints());
+
+		// Input2 consumable satisfies the constraint
+		waitUntilExecutionVertexState(ev21, ExecutionState.DEPLOYING, 2000L);
+		waitUntilExecutionVertexState(ev22, ExecutionState.DEPLOYING, 2000L);
+		ev21.getCurrentExecutionAttempt().markFinished();
+		ev22.getCurrentExecutionAttempt().markFinished();
+		assertTrue(ev31.isInputConsumable(1));
```
Comment (azagrebin) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionVertexInputConstraintTest.java]: `assertTrue(ev31.checkInputDependencyConstraints());`
instead of
`assertTrue(ev31.isInputConsumable(1));`
?

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionVertexInputConstraintTest.java [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+		ExecutionGraph eg = createExecutionGraph(ordered);
+
+		ExecutionVertex ev11 = eg.getJobVertex(v1.getID()).getTaskVertices()[0];
+		ExecutionVertex ev12 = eg.getJobVertex(v1.getID()).getTaskVertices()[1];
+		ExecutionVertex ev21 = eg.getJobVertex(v2.getID()).getTaskVertices()[0];
+		ExecutionVertex ev22 = eg.getJobVertex(v2.getID()).getTaskVertices()[1];
+		ExecutionVertex ev31 = eg.getJobVertex(v3.getID()).getTaskVertices()[0];
+		ExecutionVertex ev32 = eg.getJobVertex(v3.getID()).getTaskVertices()[1];
+
+		eg.scheduleForExecution();
```
Comment (azagrebin) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionVertexInputConstraintTest.java]: Could the first lines be part of `@Before void init()` method? They look the same for all tests.
ExecutionVertex's could be test instance variables shared by test methods.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionVertexInputConstraintTest.java [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+		// Input1 consumable does not satisfy the constraint
+		IntermediateResultPartition partition11 = ev11.getProducedPartitions().values().iterator().next();
+		ev11.scheduleOrUpdateConsumers(new ResultPartitionID(partition11.getPartitionId(),
+			ev11.getCurrentExecutionAttempt().getAttemptId()));
+		assertFalse(ev31.checkInputDependencyConstraints());
+
+		// Input2 consumable satisfies the constraint
+		ev21.getCurrentExecutionAttempt().markFinished();
+		ev22.getCurrentExecutionAttempt().markFinished();
+		assertTrue(ev31.isInputConsumable(1));
```
Comment (azagrebin) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionVertexInputConstraintTest.java]: Also here
assertTrue(ev31.checkInputDependencyConstraints());
instead of
assertTrue(ev31.isInputConsumable(1));
?

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/IntermediateResultPartitionTest.java [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+		ExecutionJobVertex jobVertex = getExecutionVertex(new JobVertexID(), new DirectScheduledExecutorService());
+		IntermediateResult result =
+			new IntermediateResult(new IntermediateDataSetID(), jobVertex, 2, ResultPartitionType.PIPELINED);
+		ExecutionVertex vertex1 =
+			new ExecutionVertex(jobVertex, 0, new IntermediateResult[]{result}, Time.minutes(1));
+		ExecutionVertex vertex2 =
+			new ExecutionVertex(jobVertex, 1, new IntermediateResult[]{result}, Time.minutes(1));
+
+		IntermediateResultPartition partition1 = result.getPartitions()[0];
+		IntermediateResultPartition partition2 = result.getPartitions()[1];
```
Comment (azagrebin) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/IntermediateResultPartitionTest.java]: Also here, the first lines could returned from a separate method `IntermediateResultPartition[] createIntermediateResultPartition()` and shared by tests.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java [hunk 10]
```diff
@@ -755,23 +775,10 @@ else if (numConsumers == 0) {
...
-								Collections.emptySet());
-						} catch (Throwable t) {
-							consumerVertex.fail(new IllegalStateException("Could not schedule consumer " +
-									"vertex " + consumerVertex, t));
-						}
-
-						return null;
-					},
-					executor);
+				if (consumerVertex.checkInputDependencyConstraints()) {
```
[the remark this answers is not in the document]
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java]: From my understanding, the TODO comment is related to the "consumerState == CREATED" section in scheduleOrUpdateConsumers, which invokes cachePartitionInfo first and then schedules the vertex. The cachePartitionInfo action is used to avoid deployment race at the cost of redundant partition infos to update to task, which is the concern as described in the TODO comment.

So far the redundant partition it's not a big problem. But I think we can optimize it later. One possible solution in my mind is to remove known partition infos from the cache when creating InputChannelDeploymentDescriptor.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/IntermediateResultPartitionTest.java (hunk 9 above)
Comment (zhuzhurk) [reply flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/IntermediateResultPartitionTest.java]: Agree.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionVertexInputConstraintTest.java (hunk 8 above)
Comment (zhuzhurk) [reply flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionVertexInputConstraintTest.java]: My fault.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionVertexInputConstraintTest.java (hunk 6 above)
Comment (zhuzhurk) [reply flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionVertexInputConstraintTest.java]: Yes. My fault.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionVertexInputConstraintTest.java (hunk 7 above)
Comment (zhuzhurk) [reply flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionVertexInputConstraintTest.java]: I will move the shared logics to a common method. As the ExecutionGraph states is changing so I'd like each test maintain its own EG.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java]: We can skip the input constraint check for `InputDependencyConstraint.ANY` because when invoking `scheduleOrUpdateConsumers`, at least one of the consumer vertex's inputs is consumable. This assumption may not be right if `checkInputDependencyConstraints` is invoked from other entries.
That's why I think it's better to add the shortcut in `scheduleOrUpdateConsumers`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionVertexInputConstraintTest.java (hunk 7 above)
Comment (azagrebin) [reply flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionVertexInputConstraintTest.java]: Yes, I meant `@Before void init()` method could recreate new graph before each test run.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java [hunk 11]
```diff
@@ -726,6 +730,41 @@ void sendPartitionInfos() {
...
+	}
+
+	/**
+	 * An input is consumable when
+	 * 1. the source result is PIPELINED and one of the result partition has produced data.
+	 * 2. the source result is BLOCKING and is FINISHED(all partitions are FINISHED).
+	 *
+	 * @return whether the input is consumable
+	 */
+	public boolean isInputConsumable(int inputNumber) {
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java]: I think this method should be moved into the `IntermediateResult`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java [hunk 12]
```diff
@@ -726,6 +730,41 @@ void sendPartitionInfos() {
...
 
+	/**
+	 * Check whether the InputDependencyConstraint is satisfied for this vertex.
+	 *
+	 * @return whether the input constraint is satisfied
+	 */
+	public boolean checkInputDependencyConstraints() {
+		if (getExecutionGraph().getInputDependencyConstraint() == InputDependencyConstraint.ANY) {
+			// InputDependencyConstraint == ANY
+			return IntStream.range(0, inputEdges.length).anyMatch(this::isInputConsumable);
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java]: Having moved `isInputConsumable` into `IntermediateResult` allows to get rid of the indirection of the `IntStream`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java [hunk 13]
```diff
@@ -726,6 +730,41 @@ void sendPartitionInfos() {
 		}
 	}
 
+	/**
+	 * Check whether the InputDependencyConstraint is satisfied for this vertex.
+	 *
+	 * @return whether the input constraint is satisfied
+	 */
+	public boolean checkInputDependencyConstraints() {
+		if (getExecutionGraph().getInputDependencyConstraint() == InputDependencyConstraint.ANY) {
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java]: Shouldn't the `InputDependencyConstraint` rather be a value of the `ExecutionJobVertex` than the `ExecutionGraph`? I guess it should be configurable for each operator individually.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java [hunk 14]
```diff
@@ -686,6 +689,7 @@ void scheduleOrUpdateConsumers(ResultPartitionID partitionId) {
 
 		if (partition.getIntermediateResult().getResultType().isPipelined()) {
 			// Schedule or update receivers of this partition
+			partition.markSomePipelinedDataProduced();
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java]: could this be generalized by having a `partition#markDataProduced` and then calling it fall intermediate results independent of the type?

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraphBuilder.java [hunk 15]
```diff
@@ -179,6 +179,13 @@ public static ExecutionGraph buildGraph(
 		executionGraph.setScheduleMode(jobGraph.getScheduleMode());
 		executionGraph.setQueuedSchedulingAllowed(jobGraph.getAllowQueuedScheduling());
 
+		try {
+			executionGraph.setInputDependencyConstraint(
+                jobGraph.getSerializedExecutionConfig().deserializeValue(classLoader).getInputDependencyConstraint());
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraphBuilder.java]: I understand that this was the easiest way how to get the `InputDependencyConstraint` into the `ExecutionGraph` but I think it should be part of the `JobVertex`, because it is not a global setting but rather controls how each vertex is scheduled.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java (hunk 11 above)
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java]: A vertex `Input` is a bit different from its corresponding `IntermediateResult` with `POINTWISE` edge. So we need the `inputEdges` info in ExecutionVertex.

I change it a bit to be more concise here: An input is consumable when when any partition in it is consumable. (whether a partition is consumable is different for PIPELINED and BLOCKING results)

B.T.W I'm also thinking about a later improvement that we can decide whether the input is consumable according to its completeness percentage. It's also a configuration related to `inputEdges`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java (hunk 14 above)
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java]: Sure.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraphBuilder.java (hunk 15 above)
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraphBuilder.java]: Good suggestion. I've move `InputDependencyConstraint ` to JobVertex.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java (hunk 13 above)
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java]: I've move `InputDependencyConstraint ` to JobVertex. And the job wide default value can be configured in `ExecutionConfig`. But I haven't make it configurable through DataSet/DataStream API yet.

I agree we should support the constraint configurable for each operator. But I'm not quite sure whether we should support it with DataSet API or later for the unified StreamGraph/Transformation API? Could you share your suggestion?

In our production experience, a job-wide configured input constraint satisfies most users, together with `BATCH_FORCED` execution mode, to ensure a batch job can finish with limited resources.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java (hunk 11 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java]: True, you're right.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java [hunk 16]
```diff
@@ -689,7 +693,7 @@ void scheduleOrUpdateConsumers(ResultPartitionID partitionId) {
 
 		if (partition.getIntermediateResult().getResultType().isPipelined()) {
 			// Schedule or update receivers of this partition
-			partition.markSomePipelinedDataProduced();
+			partition.markDataProduced();
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java]: Shouldn't we always call `partition.markDataProduced` also if it's a `BLOCKING` result partition?

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java [hunk 17]
```diff
@@ -746,23 +750,16 @@ public boolean checkInputDependencyConstraints() {
...
-	 * 1. the source result is PIPELINED and one of the result partition has produced data.
-	 * 2. the source result is BLOCKING and is FINISHED(all partitions are FINISHED).
+	 * Get whether an input of the vertex is consumable.
+	 * An input is consumable when when any partition in it is consumable.
+	 *
+	 * Note that a BLOCKING result partition is only consumable when all partitions in the result are FINISHED.
 	 *
 	 * @return whether the input is consumable
 	 */
 	public boolean isInputConsumable(int inputNumber) {
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java]: Can be package private.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java [hunk 18]
```diff
@@ -736,7 +740,7 @@ void sendPartitionInfos() {
 	 * @return whether the input constraint is satisfied
 	 */
 	public boolean checkInputDependencyConstraints() {
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java]: Can be package private.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java (hunk 16 above)
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java]: Sure,  `partition.markDataProduced` is fine for BLOCKING result. I moved it out from the `if` clause.

In my understanding, the `ExecutionVertex.scheduleOrUpdateConsumers` is for PIPELINED partition currently. Only PIPELINED partition will send the `scheduleOrUpdateConsumers` message and the method will throw exception if the partition is BLOCKING.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java (hunk 18 above)
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java]: Changed it to default scope as it will be invoked by tests.

**Comment by tillrohrmann** [comment]
Thanks for addressing my comments @zhuzhurk. Looks really good now. Merging this PR now.

**Comment by zhuzhurk** [comment]
Thanks Andrey(@azagrebin) and Till(@tillrohrmann) for the reviewing.
